### ISSUE TITLE
Slow performance on processing WBEM response

### ISSUE THREAD

**Issue by hfanan** [pywbem/pywbem#1630]
A request to get ~120K objects  from a WBEM server takes more than 10 minutes.
Time to get response from the server is about 10 seconds only.
Using cProfiler, most of the time is spent in processing xml response.

[smi-s.pstat.gz](https://github.com/pywbem/pywbem/files/2789593/smi-s.pstat.gz)

**Comment by KSchopmeyer** [comment]
The above file is a pycharm generated output and we will need pycharm to view it.  They appear to be tying up most of their time in check_node but the profile will help clarify that.

I don't have any idea what the size of the response objects is but the overall count is in the 50,000 plus instances.

One way to test this against pegasus which has a test provider that will generate a defined number of instances. and size of each instance.  We did this a long time ago and there may be a limited version of that in the run_cimoperations.py but I am not sure.

**Comment by KSchopmeyer** [comment]
More info.  They are testing with the pull also but with the pull set to pull 100,000 objects in a pull which means that the pull is not really doing anything.  They commented that with the pull set to a much lower limit they do not have the same limit.  I.e. processing enormous lists is big issue.

**Comment by andy-maier** [comment]
We agreed that we would want to analyze where pywbem spends the time in that case. Karl proposed to use his OpenPegasus test provider that can generate large artificual responses, and if the duration issue is reproducable with the pywbem client there, we can capture the response XML to create a pywbem testclient/function testcase.

**Comment by KSchopmeyer** [comment]
We expanded the performance test manual test script to capture more detailed information.  The results of that test running against a local openpegasus server are that almost all of the time is in

the TupleParser

``'
        tp = TupleParser(self.conn_id)
        tup_tree = tp.parse_cim(tt_)

The folloowing is a table showing the timing for various parts of the response and with varying number if instances in an enumerateinstances response, varying size of the pull and varying size of each response instance.

There are no major differences between enumerateinstances and the open/pull sequences or with varying sizes of the number of objects in each pull (assuming that the pull  sizes are logical (ex. between 100 and a few thousand)


'''

pp tests/manualtest/run_enum_performance.py http://localhost --response-count 1000 --pull-size 100 --response-size 100
Test with response sizes=[100] response_counts=[1000], pull_sizes=[100]

+-------------+------------+------------+-------------+----------+----------------+------------+------------+-------------+--------------+--------------+------------+---------------+----------------+-------------+
| Operation   |   Response |   RespSize | MaxObjCnt   | Result   | Exec time      |   inst/sec |   svr-time |   resp-time |   other_proc |   other_proc |   xmltuple |   tupleparser |   tuplechecker |   rtnscheck |
|             |      Count |      Bytes | Request     | Count    |                |            |        sec |         sec |          sec |      percent |        sec |           sec |            sec |         sec |
+=============+============+============+=============+==========+================+============+============+=============+==============+==============+============+===============+================+=============+
| Enum        |       1000 |        100 |             |          | 0:00:01.356963 |    736.94  |   0.004288 |    1.34515  |     1.34086  |      99.6812 |   0.158722 |       1.149   |    3.33786e-06 | 1.19209e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+-------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |       1000 |        100 | 100         | 10       | 0:00:01.482852 |    674.376 |   0.003952 |    0.149841 |     0.145889 |      97.3625 |   0.245456 |       1.05784 |    2.6e-05     | 1.5e-05     |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+-------------+--------------+--------------+------------+---------------+----------------+-------------+
(pywbem3) kschopmeyer@sheldon:~/pywbem/githubpywbem/pywbem(ks/#1630pullresponse)$ pp tests/manualtest/run_enum_performance.py http://localhost --response-count 1000 10000 50000 100000 --pull-size 100 1000 10000 --response-size 100 10000
Test with response sizes=[100, 10000] response_counts=[1000, 10000, 50000, 100000], pull_sizes=[100, 1000, 10000]

+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Operation   |   Response |   RespSize | MaxObjCnt   | Result   | Exec time      |   inst/sec |   svr-time |    resp-time |   other_proc |   other_proc |   xmltuple |   tupleparser |   tuplechecker |   rtnscheck |
|             |      Count |      Bytes | Request     | Count    |                |            |        sec |          sec |          sec |      percent |        sec |           sec |            sec |         sec |
+=============+============+============+=============+==========+================+============+============+==============+==============+==============+============+===============+================+=============+
| Enum        |       1000 |        100 |             |          | 0:00:01.334566 |    749.307 |   0.005429 |   1.3229     |    1.31747   |      99.5896 |   0.16298  |       1.12256 |    4.05312e-06 | 1.43051e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |       1000 |        100 | 10000       | 2        | 0:00:01.529491 |    653.812 |   0.002801 |   1.33352    |    1.33072   |      99.79   |   0.20333  |       1.1187  |    4e-06       | 2e-06       |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |      10000 |        100 |             |          | 0:00:14.463359 |    691.402 |   0.003484 |  14.3629     |   14.3594    |      99.9757 |   1.84849  |      12.2393  |    4.29153e-06 | 1.90735e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |      10000 |        100 | 10000       | 11       | 0:00:15.977668 |    625.874 |   0.026588 |   0.00411892 |   -0.0224691 |    -545.509  |   2.70958  |      12.8467  |    3.4e-05     | 1.8e-05     |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |      50000 |        100 |             |          | 0:01:12.270526 |    691.845 |   0.002845 |  71.752      |   71.7492    |      99.996  |   9.87358  |      60.5156  |    4.76837e-06 | 2.14577e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |      50000 |        100 | 10000       | 47       | 0:01:19.683204 |    627.485 |   0.129082 |   0.566531   |    0.437449  |      77.2154 |  12.937    |      65.1435  |    0.000172    | 9e-05       |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |     100000 |        100 |             |          | 0:02:26.263142 |    683.699 |   0.005527 | 145.183      |  145.178     |      99.9962 |  19.3342   |     123.107   |    5.72205e-06 | 2.86102e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |     100000 |        100 | 10000       | 92       | 0:02:40.249192 |    624.028 |   0.216014 |   1.31293    |    1.09692   |      83.5472 |  30.6591   |     126.442   |    0.000345    | 0.000179    |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |       1000 |      10000 |             |          | 0:00:01.516045 |    659.611 |   0.026981 |   1.50463    |    1.47765   |      98.2068 |   0.19081  |       1.06338 |    4.29153e-06 | 1.43051e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |       1000 |      10000 | 10000       | 3        | 0:00:01.554490 |    643.298 |   0.006486 |   0.169041   |    0.162555  |      96.1631 |   0.229882 |       1.05647 |    6e-06       | 4e-06       |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |      10000 |      10000 |             |          | 0:00:15.177746 |    658.859 |   0.026111 |  15.0703     |   15.0442    |      99.8267 |   1.90039  |      10.9012  |    3.8147e-06  | 1.66893e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |      10000 |      10000 | 10000       | 11       | 0:00:16.891272 |    592.022 |   0.099671 |   0.454846   |    0.355175  |      78.0869 |   3.18361  |      12.8095  |    3.7e-05     | 1.8e-05     |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |      50000 |      10000 |             |          | 0:01:23.817854 |    596.532 |   0.02453  |  83.2415     |   83.217     |      99.9705 |  11.4409   |      60.5075  |    5.24521e-06 | 1.90735e-06 |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |      50000 |      10000 | 10000       | 47       | 0:01:26.619824 |    577.235 |   0.511093 |   2.25089    |    1.7398    |      77.2937 |  17.1147   |      65.8273  |    0.000169    | 9.3e-05     |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Enum        |     100000 |      10000 |             |          | 0:02:46.805317 |    599.501 |   0.028469 | 165.791      |  165.762     |      99.9828 |  22.4038   |     121.734   |    5.48363e-06 | 2.6226e-06  |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+
| Open/Pull   |     100000 |      10000 | 10000       | 92       | 0:02:54.192621 |    574.077 |   1.08878  |   1.55791    |    0.469122  |      30.1124 |  37.8466   |     129.219   |    0.000436    | 0.000183    |
+-------------+------------+------------+-------------+----------+----------------+------------+------------+--------------+--------------+--------------+------------+---------------+----------------+-------------+

Code added to execute this was:

```

Instrumented code from _imethodcall

        self._last_xmltuple = None
        self._last_tupleparser = None
        self._last_tuplechecker = None
        self._last_rtnscheck = None
        if self.debug:
            self._last_request = req_xml.toprettyxml(indent='  ')
            self._last_reply = None


        # Send request and receive response
        reply_data, self._last_server_response_time = wbem_request(
            self.url, request_data, self.creds, cimxml_headers,
            x509=self.x509,
            verify_callback=self.verify_callback,
            ca_certs=self.ca_certs,
            no_verification=self.no_verification,
            timeout=self.timeout,
            debug=self.debug,
            recorders=self._operation_recorders,
            conn_id=self.conn_id)

        start_time = time.time()

        # Set attributes recording the response, part 1.
        # Only those that can be done without parsing (which can fail).
        self._last_raw_reply = reply_data
        self._last_reply_len = len(reply_data)

        # Parse the XML into a tuple tree (may raise CIMXMLParseError or
        # XMLParseError):
        tt_ = xml_to_tupletree_sax(reply_data, "CIM-XML response")

        self._last_xmltuple = time.time() - start_time         <--------------------------------
        start_time = time.time()

        tp = TupleParser(self.conn_id)
        tup_tree = tp.parse_cim(tt_)

        self._last_tupleparser = time.time() - start_time          <--------------------------------
        start_time = time.time()

        # Set attributes recording the response, part 2.
        if self.debug:
            self._last_reply = _to_pretty_xml(reply_data)

        # Check the tuple tree

        if tup_tree[0] != 'CIM':
            raise CIMXMLParseError(
                _format("Expecting CIM element, got {0}", tup_tree[0]),
                conn_id=self.conn_id)
        tup_tree = tup_tree[2]

        if tup_tree[0] != 'MESSAGE':
            raise CIMXMLParseError(
                _format("Expecting MESSAGE element, got {0}", tup_tree[0]),
                conn_id=self.conn_id)
        tup_tree = tup_tree[2]

        if tup_tree[0] != 'SIMPLERSP':
            raise CIMXMLParseError(
                _format("Expecting SIMPLERSP element, got {0}", tup_tree[0]),
                conn_id=self.conn_id)
        tup_tree = tup_tree[2]

        if tup_tree[0] != 'IMETHODRESPONSE':
            raise CIMXMLParseError(
                _format("Expecting IMETHODRESPONSE element, got {0}",
                        tup_tree[0]),
                conn_id=self.conn_id)

        if tup_tree[1]['NAME'] != methodname:
            raise CIMXMLParseError(
                _format("Expecting attribute NAME={0!A}, got {1!A}",
                        methodname, tup_tree[1]['NAME']),
                conn_id=self.conn_id)
        tup_tree = tup_tree[2]

        self._last_tuplechecker = time.time() - start_time         <--------------------------------
        start_time = time.time()

        # At this point tup_tree is a list of the child elements of
        # IMETHODRESPONSE:
        #   (ERROR | (IRETURNVALUE?, PARAMVALUE*)
        # or an empty list if there were no child elements.
        #
        # More specifically, the following cases are possible for tup_tree:
        # - operation failed: List with one ERROR node
        # - operation succeeded:
        #   - operation has void return type and no output parms
        #     (e.g. DeleteInstance): Empty list.
        #   - operation has non-void return type and no output parms
        #     (e.g. CreateInstance): List with (zero or?) one IRETURNVALUE
        #     node.
        #   - operation has non-void return type and output parms
        #     (e.g. OpenEnumerateInstances): List with (zero or?) one
        #     IRETURNVALUE node and one unpacked PARAMVALUE node for each
        #     output parameter.
        #
        # Note: DSP0200 is not entirely clear as to whether an empty result
        # list must be represented as an empty IRETURNVALUE node vs.
        # omitting it. Pywbem tolerates a missing IRETURNVALUE node.
        #
        # Note that there are no operations defined with a void return type but
        # with output parameters.

        # Check for failed operation
        if tup_tree and tup_tree[0][0] == 'ERROR':
            # The operation failed
            err = tup_tree[0]
            code = int(err[1]['CODE'])
            err_insts = err[2] or None  # List of CIMInstance objects
            if 'DESCRIPTION' in err[1]:
                desc = err[1]['DESCRIPTION']
            else:
                desc = _format("Error code {0}", err[1]['CODE'])
            self._last_rtnscheck = time.time() - start_time
            raise CIMError(
                code, desc, instances=err_insts, conn_id=self.conn_id)

        # At this point, we know the operation was successful.

        return_value = False
        out_param_names = []
        for child_node in tup_tree:
            if child_node[0] == 'IRETURNVALUE':
                return_value = True
            else:
                # The PARAMVALUE nodes are already unpacked
                out_param_names.append(child_node[0])

        # Convert empty list to None
        if not tup_tree:
            tup_tree = None

        # Check unexpected return value and output parameters
        if not has_return_value and return_value:
            self._last_rtnscheck = time.time() - start_time
            raise CIMXMLParseError(
                _format("Unexpected IRETURNVALUE child element of "
                        "IMETHODRESPONSE for operation {0} which is defined as "
                        "void", methodname),
                conn_id=self.conn_id)
        if not has_out_params and out_param_names:
            self._last_rtnscheck = time.time() - start_time
            raise CIMXMLParseError(
                _format("Unexpected PARAMVALUE child element(s) of "
                        "IMETHODRESPONSE for operation {0} which has no output "
                        "parameters defined: {1!A}",
                        methodname, out_param_names),
                conn_id=self.conn_id)

        # Missing return value and output parameters are checked by caller
        self._last_rtnscheck = time.time() - start_time         <--------------------------------

        return tup_tree

**Comment by andy-maier** [comment]
Fixed by PR #1658
